include/xrpl/protocol/Asset.h

```diff
@@ -0,0 +1,174 @@
+ * For historical reasons, two issue types XRP and IOU are wrapped in Issue
+ * type. Asset replaces Issue where any issue type is expected. For instance,
+ * STAmount replaces Issue with Asset to represent any issue amount.
+ */
+class Asset
+{
+private:
+    using value_type = std::variant<Issue, MPTIssue>;
+    value_type issue_;
+
+public:
+    Asset() = default;
+
+    /** Conversions to Asset are implicit and conversions to specific issue
+     *  type are explicit. This design facilitates the use of Asset.
+     */
+    Asset(Issue const& issue) : issue_(issue)
+    {
+    }
+
+    Asset(MPTIssue const& mptIssue) : issue_(mptIssue)
+    {
+    }
+
+    Asset(MPTID const& issuanceID) : issue_(MPTIssue{issuanceID})
+    {
+    }
+
+    AccountID const&
+    getIssuer() const;
+
+    template <ValidIssueType TIss>
+    constexpr TIss const&
+    get() const;
+
+    template <ValidIssueType TIss>
+    TIss&
+    get();
+
+    template <ValidIssueType TIss>
+    constexpr bool
+    holds() const;
+
+    std::string
+    getText() const;
+
+    constexpr value_type const&
+    value() const;
+
+    void
+    setJson(Json::Value& jv) const;
+
+    bool
+    native() const
+    {
+        return holds<Issue>() && get<Issue>().native();
+    }
+
+    friend constexpr bool
+    operator==(Asset const& lhs, Asset const& rhs);
+
+    friend constexpr bool
+    operator!=(Asset const& lhs, Asset const& rhs);
+
+    friend constexpr bool
+    operator==(Currency const& lhs, Asset const& rhs);
+};
+
+template <ValidIssueType TIss>
+constexpr bool
+Asset::holds() const
+{
+    return std::holds_alternative<TIss>(issue_);
+}
+
+template <ValidIssueType TIss>
+constexpr TIss const&
+Asset::get() const
+{
+    if (!std::holds_alternative<TIss>(issue_))
+        Throw<std::logic_error>("Asset is not a requested issue");
+    return std::get<TIss>(issue_);
+}
+
+template <ValidIssueType TIss>
+TIss&
+Asset::get()
+{
+    if (!std::holds_alternative<TIss>(issue_))
+        Throw<std::logic_error>("Asset is not a requested issue");
+    return std::get<TIss>(issue_);
+}
+
+constexpr Asset::value_type const&
+Asset::value() const
+{
+    return issue_;
+}
+
+constexpr bool
+operator==(Asset const& lhs, Asset const& rhs)
+{
+    return std::visit(
+        [&]<typename TLhs, typename TRhs>(
+            TLhs const& issLhs, TRhs const& issRhs) {
+            if constexpr (std::is_same_v<TLhs, TRhs>)
+                return issLhs == issRhs;
+            else
+                return false;
+        },
+        lhs.issue_,
+        rhs.issue_);
+}
+
+constexpr bool
+operator!=(Asset const& lhs, Asset const& rhs)
+{
+    return !(lhs == rhs);
+}
+
+constexpr bool
+operator==(Currency const& lhs, Asset const& rhs)
+{
+    return rhs.holds<Issue>() && rhs.get<Issue>().currency == lhs;
+}
+
+inline bool
+isXRP(Asset const& asset)
+{
+    return asset.native();
+}
+
+std::string
+to_string(Asset const& asset);
+
+bool
+validJSONAsset(Json::Value const& jv);
```

Comment:
Can you please write an `assetFromJson`? You may want to delegate to `issueFromJson` (which exists) and `mptIssueFromJson` (which does not exist). I need in the `stissue` branch, in `STParsedJSON.cpp`, in the case for `STI_ISSUE` (which right now calls `issueFromJson`).